examples/counter/tests/single_chain.rs

```diff
@@ -18,7 +18,7 @@ async fn single_chain_test() {
 
     let initial_state = 42u64;
     let application_id = chain
-        .create_application::<counter::CounterAbi>(bytecode_id, (), initial_state, vec![])
+        .create_application::<counter::CounterAbi, (), u64>(bytecode_id, (), initial_state, vec![])
```

Comment:
This is hurting type-safety. A possible solution would be to write `create_application::<counter::CounterAbi, counter::Parameters, counter::InitializationArgument>` or `create_application::<counter::App>` for a new type (and trait) `App`.

@afck @jvff What do you think?

Reply:
Right, that's a good point.

Basically it's even the `bytecode_id` that's only compatible with specific initialization arguments and parameters. I'm wondering whether we should make `BytecodeId<Abi, InitializationArgument, Parameters>` generic in all three, and then leave the `create_application` parameters implicit.

Reply:
`BytecodeId<Abi, InitializationArgument, Parameters>` makes sense to me

We may eventually have to rename `InitializationArgument` into `InstantiationArgument` and `initialize` into `instantiate` by the way.

Reply:
I'm not sure I get the correlation of `BytecodeId` with `InitializationArgument` and `Parameters`, could one of you explain this with more detail? And maybe some code pointers. From what I looked you need those just to create the application, `BytecodeId` doesn't use or need neither

Reply:
Your proposal works but if you write `u32` instead of `u64`, the Rust type-checker won't complain.

Reply:
Ok, sounds good, I can do that. Shouldn't CI be red though if there's a type checker issue? 🤔

Reply:
What I mean is that it doesn't make sense to call `create_application` with the same bytecode ID but different initialization argument or parameter types: It's the same code, so it expects the same types.

That's why I was thinking we could make them arguments of `BytecodeId` (but not `ApplicationId`, because once an application is created, they don't matter anymore). Then the compiler wouldn't accept if you called `create_application` with different types.

Reply:
(But not a blocker from my side, just a thought.)